File: app/ios/NeuroPulse/Models/AdaptationEvent.swift
import Foundation

// Mirrors np_adapt_trigger_t in firmware/hub_control/include/np_adaptation_log.h
// Plain-language copy requires Privacy Lead sign-off (OI-PA-04) before shipping.
// Never add raw EEG or HRV float values to this file — trigger context is enum-only.

enum AdaptTrigger: UInt8, CaseIterable {
    // EEG band-power triggers
    case eegAlphaLow        = 0x00
    case eegAlphaHigh       = 0x01
    case eegThetaLow        = 0x02
    case eegThetaHigh       = 0x03
    case eegGammaLow        = 0x04
    case eegGammaHigh       = 0x05
    case eegBetaLow         = 0x06
    case eegBetaHigh        = 0x07
    // HRV triggers
    case hrvCoherenceLow    = 0x08
    case hrvCoherenceHigh   = 0x09
    case hrvRmssdLow        = 0x0A
    // Safety / device-condition triggers
    case impedanceChange    = 0x0B
    case doseLimitApproach  = 0x0C
    case thermalThrottle    = 0x0D
    // Session-lifecycle triggers
    case phaseTransition    = 0x0E
    case userPause          = 0x0F
    case sessionEndRamp     = 0x10

    // Approved plain-language descriptions — extend only with Privacy Lead sign-off (OI-PA-04).
    var plainLanguageDescription: String {
        switch self {
        case .eegAlphaLow:
            return "Brainwave activity in the calm-focus range dipped — stimulation frequency was increased to support focus."
        case .eegAlphaHigh:
            return "Brainwave activity in the calm-focus range was elevated — stimulation frequency was reduced to avoid over-stimulation."
        case .eegThetaLow:
            return "Brainwave activity in the memory-processing range dipped — stimulation was adjusted to support memory encoding."
        case .eegThetaHigh:
            return "Brainwave activity in the memory-processing range was elevated — stimulation was reduced."
        case .eegGammaLow:
            return "High-frequency brainwave activity (linked to focus and attention) dipped — stimulation frequency was stepped up."
        case .eegGammaHigh:
            return "High-frequency brainwave activity was elevated — stimulation frequency was stepped down to prevent fatigue."
        case .eegBetaLow:
            return "Alertness-linked brainwave activity dipped — stimulation was adjusted to maintain engagement."
        case .eegBetaHigh:
            return "Alertness-linked brainwave activity was elevated — stimulation was reduced to encourage relaxation."
        case .hrvCoherenceLow:
            // swiftlint:disable:next line_length
            return "Heart rhythm coherence fell below your personal target — the breathing pacer rate was adjusted to help restore coherence."
        case .hrvCoherenceHigh:
            // swiftlint:disable:next line_length
            return "Heart rhythm coherence was above target — the breathing pacer rate was eased to maintain the effect without over-correction."
        case .hrvRmssdLow:
            return "Heart rate variability fell below your baseline — stimulation intensity was reduced to lower physiological load."
        case .impedanceChange:
            return "Electrode contact quality changed during the session — stimulation was adjusted to account for the change."
        case .doseLimitApproach:
            return "The session was approaching its cumulative light-energy limit — output was reduced to stay within the safe dose."
        case .thermalThrottle:
            return "A device component reached its temperature limit — output was reduced temporarily to allow cooling."
        case .phaseTransition:
            return "The session moved to a new phase — stimulation parameters were updated for the new phase."
        case .userPause:
            return "The session was paused — all stimulation was reduced to standby levels."
        case .sessionEndRamp:
            return "The session was completing — all stimulation was gradually reduced to zero."
        }
    }
}

// One closed-loop adaptation event as decoded from the UHDR log.
// Mirrors np_adaptation_event_t — no raw biometric floats.
struct AdaptationEvent: Identifiable {
    // Stable identity derived from content — the same UHDR event decoded twice
    // produces the same id, so SwiftUI diffing is correct across re-renders.
    // Format: "<sessionOffsetMs>-<trigger.rawValue>-<modType>-<paramId>"
    var id: String { "\(sessionOffsetMs)-\(trigger.rawValue)-\(modType)-\(paramId)" }

    var sessionOffsetMs:  UInt32
    var trigger:          AdaptTrigger
    var modType:          UInt8       // np_hub_mod_type_t raw value
    var paramId:          UInt8
    var valueBeforeX100:  Int32       // parameter value before × 100
    var valueAfterX100:   Int32       // parameter value after × 100
    var confidencePct:    UInt8       // 0–100; 0xFF = not applicable

    var sessionOffset: TimeInterval { TimeInterval(sessionOffsetMs) / 1000.0 }
}

// Post-session summary handed to SessionHistoryView.
//
// Display-aggregated metrics only — no raw EEG/HRV waveform data. The optional
// coherence/RMSSD/impedance fields are the same values shown live during the
// session and are what SessionHistoryStore persists.
struct CompletedSessionSummary {
    var protocolName:      String
    var durationSeconds:   UInt32
    var adaptationEvents:  [AdaptationEvent]
    var completedAt:       Date

    var averageCoherenceScore: Float?  = nil   // 0.0–10.0; nil when no HRV modality
    var rmssdMilliseconds:     UInt16? = nil    // integer only
    var impedancePassCount:    Int     = 0      // 0–8 EEG electrodes passed
    var edfSessionID:          UInt32? = nil    // hub session ID for Mode 4 EDF download

    // Reconstruct a summary from a persisted history row. Adaptation events are
    // not persisted in the history store (they live in UHDR), so the detail view
    // opened from history shows the empty-state Adaptive Adjustments card.
    //
    // completedAt is reconstructed from sessionDay at start-of-day in local
    // timezone. Time-of-day is not available — the store holds day granularity
    // only (UHDR boundary; NP-PRIV-ANALYSIS-002 MEDIUM-07).
    init(record: SessionRecord) {
        self.protocolName          = record.protocolName
        self.durationSeconds       = UInt32(record.durationSeconds.rounded())
        self.adaptationEvents      = []
        self.completedAt           = SessionRecord.dayFormatter.date(from: record.sessionDay) ?? .distantPast
        self.averageCoherenceScore = record.averageCoherenceScore
        self.rmssdMilliseconds     = record.rmssdMilliseconds
        self.impedancePassCount    = record.impedancePassCount
        self.edfSessionID          = record.edfSessionID
    }

    init(protocolName: String,
         durationSeconds: UInt32,
         adaptationEvents: [AdaptationEvent],
         completedAt: Date,
         averageCoherenceScore: Float? = nil,
         rmssdMilliseconds: UInt16? = nil,
         impedancePassCount: Int = 0,
         edfSessionID: UInt32? = nil) {
        self.protocolName          = protocolName
        self.durationSeconds       = durationSeconds
        self.adaptationEvents      = adaptationEvents
        self.completedAt           = completedAt
        self.averageCoherenceScore = averageCoherenceScore
        self.rmssdMilliseconds     = rmssdMilliseconds
        self.impedancePassCount    = impedancePassCount
        self.edfSessionID          = edfSessionID
    }
}

File: app/android/core/src/main/kotlin/com/neuropulse/core/session/AdaptationEvent.kt
package com.neuropulse.core.session

// Port of iOS AdaptationEvent.swift — mirrors np_adapt_trigger_t / np_adaptation_event_t in
// firmware/hub_control/include/np_adaptation_log.h. One closed-loop adaptation event decoded
// from the UHDR adapt log. PRIVACY: enum-only trigger context — never any raw EEG/HRV floats.
// Plain-language copy requires Privacy Lead sign-off (OI-PA-04) before shipping.

enum class AdaptTrigger(val rawValue: Int) {
    // EEG band-power triggers
    EEG_ALPHA_LOW(0x00),
    EEG_ALPHA_HIGH(0x01),
    EEG_THETA_LOW(0x02),
    EEG_THETA_HIGH(0x03),
    EEG_GAMMA_LOW(0x04),
    EEG_GAMMA_HIGH(0x05),
    EEG_BETA_LOW(0x06),
    EEG_BETA_HIGH(0x07),
    // HRV triggers
    HRV_COHERENCE_LOW(0x08),
    HRV_COHERENCE_HIGH(0x09),
    HRV_RMSSD_LOW(0x0A),
    // Safety / device-condition triggers
    IMPEDANCE_CHANGE(0x0B),
    DOSE_LIMIT_APPROACH(0x0C),
    THERMAL_THROTTLE(0x0D),
    // Session-lifecycle triggers
    PHASE_TRANSITION(0x0E),
    USER_PAUSE(0x0F),
    SESSION_END_RAMP(0x10);

    /** Approved plain-language copy — extend only with Privacy Lead sign-off (OI-PA-04). */
    val plainLanguageDescription: String
        get() = when (this) {
            EEG_ALPHA_LOW ->
                "Brainwave activity in the calm-focus range dipped — stimulation frequency was increased to support focus."
            EEG_ALPHA_HIGH ->
                "Brainwave activity in the calm-focus range was elevated — stimulation frequency was reduced to avoid over-stimulation."
            EEG_THETA_LOW ->
                "Brainwave activity in the memory-processing range dipped — stimulation was adjusted to support memory encoding."
            EEG_THETA_HIGH ->
                "Brainwave activity in the memory-processing range was elevated — stimulation was reduced."
            EEG_GAMMA_LOW ->
                "High-frequency brainwave activity (linked to focus and attention) dipped — stimulation frequency was stepped up."
            EEG_GAMMA_HIGH ->
                "High-frequency brainwave activity was elevated — stimulation frequency was stepped down to prevent fatigue."
            EEG_BETA_LOW ->
                "Alertness-linked brainwave activity dipped — stimulation was adjusted to maintain engagement."
            EEG_BETA_HIGH ->
                "Alertness-linked brainwave activity was elevated — stimulation was reduced to encourage relaxation."
            HRV_COHERENCE_LOW ->
                "Heart rhythm coherence fell below your personal target — the breathing pacer rate was adjusted to help restore coherence."
            HRV_COHERENCE_HIGH ->
                "Heart rhythm coherence was above target — the breathing pacer rate was eased to maintain the effect without over-correction."
            HRV_RMSSD_LOW ->
                "Heart rate variability fell below your baseline — stimulation intensity was reduced to lower physiological load."
            IMPEDANCE_CHANGE ->
                "Electrode contact quality changed during the session — stimulation was adjusted to account for the change."
            DOSE_LIMIT_APPROACH ->
                "The session was approaching its cumulative light-energy limit — output was reduced to stay within the safe dose."
            THERMAL_THROTTLE ->
                "A device component reached its temperature limit — output was reduced temporarily to allow cooling."
            PHASE_TRANSITION ->
                "The session moved to a new phase — stimulation parameters were updated for the new phase."
            USER_PAUSE ->
                "The session was paused — all stimulation was reduced to standby levels."
            SESSION_END_RAMP ->
                "The session was completing — all stimulation was gradually reduced to zero."
        }

    companion object {
        fun from(rawValue: Int): AdaptTrigger? = entries.firstOrNull { it.rawValue == rawValue }
    }
}

data class AdaptationEvent(
    val sessionOffsetMs: Long,
    val trigger: AdaptTrigger,
    val modType: Int,        // np_hub_mod_type_t raw value
    val paramId: Int,
    val valueBeforeX100: Int, // parameter value before × 100
    val valueAfterX100: Int,  // parameter value after × 100
    val confidencePct: Int,   // 0–100; 0xFF = not applicable
) {
    /** Stable content-derived identity (same UHDR event decoded twice → same id). */
    val id: String get() = "$sessionOffsetMs-${trigger.rawValue}-$modType-$paramId"

    val sessionOffsetSeconds: Double get() = sessionOffsetMs / 1000.0
}
